lib/linked_list.rb

```diff
@@ -1,45 +1,101 @@
 require_relative 'node'
+require "pry"
 
 class LinkedList
   attr_reader :head
 
   def initialize
     @head = nil
+    # why is head initially set to nil?
   end
 
-  # Time complexity - ?
-  # Space complexity - ?
+  # Time complexity - O(1) because inserting one node will never scale. It will always just be creating one node. 
+  # Space complexity - O(1) because inserting one node will always use one location in memory
   def add_first(data)
-
+    new_node = Node.new(data)
+    new_node.next = @head
+    # Need to put the node to another node or else it'll just
+    # be a floating node
+    @head = new_node
   end
 
-  # Time complexity - ?
-  # Space complexity - ?
+  # Time complexity - O(1) because searching for the head node node will never scale. 
+  # Space complexity - O(1) because searching for the first node will never scale.
+  # a little confused as to how the space complexity is 0(1) though. We aren't storing anything.
   def get_first
+    if @head == nil
+      return nil
+    end 
 
+    @head.data
   end
 
-  # Time complexity - ?
-  # Space complexity - ?
+  # Time complexity - O(n) because you have to traverse the entire list to find the length so the time complexity
+  # will scale according to the length of the list
+  # Space complexity - O(1) because we will never use more than one variable (memory) to traverse the linked list.
+
   def length
-    return 0
+    
+    current = @head
+
+    count = 0
+
+    while current != nil 
+      
+      count += 1
+      current = current.next
+    end 
+
+    return count 
   end
 
-  # Time complexity - ?
-  # Space complexity - ?
+  # Time complexity - 0(n) because we have to traverse the whole list to add to the last node. To access the nth
+  # element of any linked list, you must traverse it.
+  # Space complexity - 0(1) because only one node will be created at a time
+  # or is it because we will only ever use one variable?
   def add_last(data)
+    current = @head 
 
+    if current == nil
+      return self.add_first(data)
+    end
+
+    while current.next != nil
+      current = current.next
+    end
+    
+    current.next = Node.new(data)
+    
   end
 
-  # Time complexity - ?
-  # Space complexity - ?
+  # Time complexity - O(n) because we have to traverse the whole list to add to the last node. To access the nth
+  # element of any linked list, you must traverse it.
+  # Space complexity - 0(1) because we will only ever use one variable (location in memory) to traverse the list
+  # to get the last node
   def get_last
+    current = @head
+
+    while current.next != nil
```

Comment:
What if `@head` is nil?

```suggestion
    return nil if @head.nil?
    while current.next != nil
```